client/src/components/molecules/DonationItemForm.tsx

```diff
@@ -0,0 +1,307 @@
+        }
+        setIsConditionErrorneous(false);
+        return false;
+    };
+
+    const updateIsQuantityErrorneous = (quantity: number): boolean => {
+        if (quantity <= 0 || isNaN(quantity)) {
+            setIsQuantityErrorneous(true);
+            return true;
+        }
+        setIsQuantityErrorneous(false);
+        return false;
+    };
+
+    const updateFormError = (condition: ItemCondition | null, quantity: number): string => {
+        const errors = [];
+        if (updateIsConditionErrorneous(condition)) {
+            errors.push("Item Condition");
+        }
+        if (updateIsQuantityErrorneous(quantity)) {
+            errors.push("Quantity (number of this specific item)");
+        }
+        if (errors.length === 0) {
+            setFormError("");
+            return "";
+        }
+        let errorString = "Please enter valid";
+        errors.forEach((error, idx) => {
+            if (idx === errors.length - 1 && errors.length > 1) {
+                errorString = errorString.concat(" and");
+            }
+            errorString = errorString.concat(" " + error);
+        });
+        errorString = errorString.concat(".");
+        setFormError(errorString);
+
+        return errorString;
+    };
+
+    const onConditionChange = (newCondition: ItemCondition) => {
+        setCondition(newCondition);
+        setIsConditionDropdownOpen(false);
+        updateFormError(newCondition, quantity);
+    };
+
+    const onAgeChange = (newAge: number) => {
+        setAge(newAge);
+        setIsAgeDropdownOpen(false);
+    };
+
+    const onQuantityChange = (newQuantity: number) => {
+        setQuantity(newQuantity);
+        updateFormError(condition, newQuantity);
+    };
+
+    const onDescriptionChange = (newDescription: string) => {
+        setDescription(newDescription);
+    };
+
+    const onSave = () => {
+        const nameError = updateNameError(name);
+        const formError = updateFormError(condition, quantity);
+
+        if (nameError.length === 0 && formError.length === 0) {
+            const itemRequestGroup = props.requestGroups.find((requestGroup) => requestGroup.name === name);
+
+            props.onSave({
+                age,
+                condition: condition ?? undefined,
+                description,
+                name,
+                quantity,
+                requestGroup: itemRequestGroup ?? null
+            });
+        }
+    };
+
+    return (
+        <div className="donation-item-form">
+            <FormItem
+                className="item-name-field"
+                formItemName="What item do you wish to donate?"
+                errorString={nameError}
+                isDisabled={false}
+                showErrorIcon={false}
+                inputComponent={
+                    <SearchableDropdown
+                        selectedItem={name}
+                        searchString={nameInput}
+                        dropdownItems={props.requestGroups.reduce(
```

Comment:
Just wondering if we could use map here instead of reduce?

Reply:
Yes def we could use reduce, but then we would either have to 
(1) cast `requestGroup.name` as a string (rn its type is `string|null`) 
(2) OR perform a `filter` for `null` elements after extracting request group names using `map`

(1) is bad practice (i.e. casting should be avoided if possible)... although we've been doing it in other files LOLL
(2) would require us to loop through the array twice